# 🚗 Car Price Prediction – Regression Model

This project focuses on predicting the price of a car based on its features and specifications using machine learning regression techniques.

The dataset contains information about different cars, such as brand, model, year, mileage, fuel type, transmission, engine characteristics, and other relevant features. 

The goal is to analyze these features and build a machine learning model that can estimate the expected price of a car.

#  Data Exploration

* 📊 Number of rows and columns
* 🔤 Data types
* ❌ Missing values
* 🎯 Target variable: priceUSD
* 🔢 Numerical columns
* ⚠️ Extreme values and outliers
* 📈 Data distribution and potential anomalies

# Step 1: 
* Import pandas library
* Read and load the dataset
* Info about columns

In [62]:
import pandas as pd
url = ("https://raw.githubusercontent.com/adelastrukar44-debug/car_price_prediction/main/data/cars.csv")

try:
    df = pd.read_csv(url)
    print("---- ✅Dataset loaded successfully.")
except FileNotFoundError:
    print("Error: The dataset file was not found.")
except pd.errors.EmptyDataError:
    print("Error: The dataset is empty.")
except pd.errors.ParserError:
    print("Error: Failed to parse the dataset.")
except Exception as e:
    print(f"An unexpected error occurred: {e}")

df.columns


---- ✅Dataset loaded successfully.


Index(['make', 'model', 'priceUSD', 'year', 'condition', 'mileage(kilometers)',
       'fuel_type', 'volume(cm3)', 'color', 'transmission', 'drive_unit',
       'segment'],
      dtype='str')

# Step 2: 
* 📊 Number of rows and columns
* ✅Rename some columns 

In [63]:
df.rename(columns={'mileage(kilometers)': 'mileage_km',
                  'priceUSD':'price_usd',
                 'volume(cm3)':'volume_cm3'
                 }, inplace=True)

print("Number of rows and columns:",df.shape)

Number of rows and columns: (56244, 12)


# Step 3:
 
* 🔤 Data types
* ❌ Missing values

In [64]:
print("🔍Main Info.")
print("-"*40)
print(df.info())
print("-"*40)
print("----Count Missing Values in dataset:")
print(df.isna().sum())


🔍Main Info.
----------------------------------------


<class 'pandas.DataFrame'>
RangeIndex: 56244 entries, 0 to 56243
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   make          56244 non-null  str    
 1   model         56244 non-null  str    
 2   price_usd     56244 non-null  int64  
 3   year          56244 non-null  int64  
 4   condition     56244 non-null  str    
 5   mileage_km    56244 non-null  float64
 6   fuel_type     56244 non-null  str    
 7   volume_cm3    56197 non-null  float64
 8   color         56244 non-null  str    
 9   transmission  56244 non-null  str    
 10  drive_unit    54339 non-null  str    
 11  segment       50953 non-null  str    
dtypes: float64(2), int64(2), str(8)
memory usage: 5.1 MB
None
----------------------------------------
----Count Missing Values in dataset:
make               0
model              0
price_usd          0
year               0
condition          0
mileage_km         0
fuel_type          0
volume_cm3  

### 🔍 Main Info

* 📊 The dataset contains **56,244 rows and 12 columns**.
* 🔢 There are **4 numerical columns** and **8 categorical columns**.
* 🎯 The target variable is **price_usd**.
* ❌ **volume_cm3** contains **47 missing values**.
* ❌ **drive_unit** contains **1,905 missing values**.
* ❌ **segment** contains **5,291 missing values**.
* ✅ The remaining columns have **no missing values**.
* 💾 The dataset uses approximately **5.1 MB of memory**.

### Conclusion 
Current data types are appropriate for the dataset.





# Step 4:
* 🎯 Target variable: price_usd -- check values

In [65]:
print(df['price_usd'].describe())

count     56244.000000
mean       7415.456440
std        8316.959261
min          48.000000
25%        2350.000000
50%        5350.000000
75%        9807.500000
max      235235.000000
Name: price_usd, dtype: float64


⚠️ The maximum price of $235,235 is exceptionally high compared to the median price of $5,350.
 
This value should be investigated further to determine whether it represents a genuine high-priced vehicle or a potential outlier/data anomaly.

# Step 5: Info about price_usd
* 📈 Data distribution and potential anomalies
* ⚠️ Extreme values and outliers


In [66]:
print("🔍🚗Car with the Highest Price.")
print(df.loc[df['price_usd'].idxmax()])
print("-"*40)
print("🔍🚗Car with the lowest Price.")
print(df.loc[df['price_usd'].idxmin()])

🔍🚗Car with the Highest Price.
make                    bentley
model            continental-gt
price_usd                235235
year                       2018
condition          with mileage
mileage_km              29000.0
fuel_type                petrol
volume_cm3               6000.0
color                     white
transmission               auto
drive_unit      all-wheel drive
segment                     NaN
Name: 20787, dtype: object
----------------------------------------
🔍🚗Car with the lowest Price.
make                    zaz
model                   968
price_usd                48
year                   1910
condition       with damage
mileage_km          25000.0
fuel_type            petrol
volume_cm3           1100.0
color                 green
transmission      mechanics
drive_unit       rear drive
segment                   B
Name: 5952, dtype: object


### 🔎 Maximum Price Investigation

The highest price in the dataset is **$235,235**. After investigating the corresponding record, I found that it belongs to a **2018 Bentley Continental GT** with **29,000 km** mileage and a **6000 cm³** petrol engine.

The value appears to be plausible for this type of luxury vehicle, so it should not be automatically treated as an incorrect value. However, it represents an extreme value compared to most cars in the dataset and may influence the regression model.

### 🔎 Minimum Price Investigation

The minimum price is $48 for a damaged ZAZ 968. The year 1910 appears to be anomalous, but since the correct year cannot be confirmed with certainty, the original value will be retained for now.

# Step 6: Info about mileage_km
* 📈 Data distribution and potential anomalies
* ⚠️ Extreme values and outliers


In [67]:
print("🔍🚗Car with the Highest Mileage(kilometers).")
print(df.loc[df['mileage_km'].idxmax()])
print("-"*40)
print("🔍🚗Car with the lowest Mileage(kilometers).")
print(df.loc[df['mileage_km'].idxmin()])

🔍🚗Car with the Highest Mileage(kilometers).
make                         audi
model                          80
price_usd                     500
year                         1986
condition            with mileage
mileage_km              9999999.0
fuel_type                  petrol
volume_cm3                 1800.0
color                      silver
transmission            mechanics
drive_unit      front-wheel drive
segment                         D
Name: 1394, dtype: object
----------------------------------------
🔍🚗Car with the lowest Mileage(kilometers).
make                        mazda
model                           3
price_usd                    5500
year                         2007
condition            with mileage
mileage_km                    0.0
fuel_type                  petrol
volume_cm3                 1600.0
color                      silver
transmission            mechanics
drive_unit      front-wheel drive
segment                         C
Name: 184, dtype: object


### 🔍🚗 Mileage Analysis

* 🚨 **Highest mileage:** 9,999,999 km – Audi 80 (1986). This value appears to be an extreme data anomaly and should be investigated further.
* 🔍 **Lowest mileage:** 0 km – Mazda 3 (2007). This value may be valid, but it could also represent missing or unknown mileage and should be investigated.


🔍🚗 Further Investigation -------

* Count total anomaly values.
* Display anomaly rows.

In [68]:
print("Total number of rows with mileage_km in (9999999.0).")
print((df['mileage_km'] == 9999999.0).sum())
print("Total number of rows with mileage_km in (0.0)")
print((df['mileage_km'] == 0).sum())

Total number of rows with mileage_km in (9999999.0).
22
Total number of rows with mileage_km in (0.0)
190


In [69]:
print(df[df['mileage_km'] == 9999999.0])
print("-"*40)
print(df[df['mileage_km'] == 0])

                make     model  price_usd  year     condition  mileage_km  \
1394            audi        80        500  1986  with mileage   9999999.0   
1854            audi       100        900  1989  with mileage   9999999.0   
5783         peugeot       806       4100  2000  with mileage   9999999.0   
6114        lada-vaz      2104        600  1990  with mileage   9999999.0   
6465        lada-vaz      2108        295  1990  with mileage   9999999.0   
6684             raf      2203       1000  1991  with mileage   9999999.0   
10032            bmw  5-seriya       2000  1985  with mileage   9999999.0   
10316            bmw  7-seriya       4650  1998  with mileage   9999999.0   
11699           audi        a4       3000  1996  with mileage   9999999.0   
16705           fiat     brava       1100  1997  with mileage   9999999.0   
21525          honda      cr-v       7300  2005  with mileage   9999999.0   
23161  mercedes-benz   e-klass       1850  1989  with mileage   9999999.0   

🔍🚗 Further Investigation – Extreme Mileage

After investigating all 22 vehicles with a mileage of 9,999,999 km, the value appears across different brands, models, years, prices, and vehicle conditions.
Since such a mileage value is unrealistic and occurs repeatedly across unrelated vehicles, it is most likely a placeholder or data entry anomaly rather than the actual mileage.
Therefore, these values should be treated as invalid mileage values during data preprocessing.

🔍🚗 Further Investigation – Zero Mileage

There are 190 vehicles with a mileage of 0 km. These values occur across different brands, models, years, and vehicle conditions.
Since zero mileage is unusual for used vehicles, these records require further investigation. However, there is not enough evidence to conclude that all of them are incorrect, so the values will be retained for now and considered during data preprocessing.

In [70]:
df[df['mileage_km'] == 0]['condition'].value_counts()

condition
with mileage    174
for parts        14
with damage       2
Name: count, dtype: int64

🔍🚗 Zero Mileage Conclusion

Among the 190 vehicles with 0 km mileage:

* 🚗 174 have the condition with mileage
* 🔧 14 have the condition for parts
* ⚠️ 2 have the condition with damage

The 174 vehicles listed as with mileage with 0 km are particularly unusual and may indicate missing or incorrectly recorded mileage. However, without reliable evidence of the actual mileage, these values will not be changed at this stage.

# Step 7:
* 🚗 Model Analysis



In [71]:

print(df['model'].value_counts())
print(df['model'].nunique())

model
passat      2086
5-seriya    1476
a6          1276
golf        1070
astra       1013
            ... 
xb             1
xc40           1
xjs            1
xt5            1
z3             1
Name: count, Length: 1034, dtype: int64
1034


### 🚗 Model  Analysis Conclusion

The `model` column contains **1,034 unique categories**. Some models occur frequently, while others appear only once or a few times in the dataset.

This high cardinality should be considered during preprocessing, as applying One-Hot Encoding directly would create a large number of additional features.
